### PR TITLE
Implement psa_sign_message and psa_verify_message

### LINKED ISSUE #3267: Implement psa_sign_message and psa_verify_message
Implement the PSA Crypto API 1.0.0 functions `psa_sign_messsage` and `psa_verify_message`.

Since Mbed TLS only supports hash-and-sign algorithms, these are simple wrappers around `psa_hash_compute` and `psa_sign_hash`/`psa_verify_hash`.

Also part of the goal of this task is to implement the corresponding policies `PSA_KEY_USAGE_SIGN_MESSAGE` and `PSA_KEY_USAGE_VERIFY_MESSAGE`.

### PR DESCRIPTION
**Pull request by gabor-mezei-arm**
## Description
Implement the PSA Crypto API 1.0.0 functions psa_sign_messsage and psa_verify_message.

Since Mbed TLS only supports hash-and-sign algorithms, these are simple wrappers around psa_hash_compute and psa_sign_hash/psa_verify_hash.

Also part of the goal of this task is to implement the corresponding policies PSA_KEY_USAGE_SIGN_MESSAGE and PSA_KEY_USAGE_VERIFY_MESSAGE.

Resolve #3267 

## Requires Forwardporting
- [x] mbedtls-3.0  #4510

### CODE REVIEW

File: include/psa/crypto_values.h
Comment (gilles-peskine-arm) [inline include/psa/crypto_values.h:1502]: Rather than repeat `PSA_ALG_IS_HASH_AND_SIGN`, call it. And add `|| (alg) == PSA_ALG_PURE_EDDSA` which is missing.

File: library/psa_crypto.c
Comment (gilles-peskine-arm) [inline library/psa_crypto.c:2876]: The code beyond this point is only valid when `PSA_ALG_IS_SIGN_HASH( alg )`. Otherwise the hash-then-sign method is not valid. Currently, only one algorithm is declared which is not a sign-hash algorithm: `PSA_ALG_PURE_EDDSA`. And it is not yet implemented in Mbed TLS, so it can't be mentioned here, but we must not attempt to do the hash-then-sign method for it.

Also goes for verify.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I made a first review pass, just for design. I did not review the documentation, the code or the test data in detail. I'm happy with the design. The test coverage looks ok (but I may ask for a few specific tests once I've done a detailed review).

Some things are missing. They can be done either in this PR or in a follow-up.
* [x] Drivers can have a sign/verify-message entry point. So there needs to be a `psa_driver_wrapper_sign_message` and a `psa_driver_wrapper_sign_hash`. If there is no all-in-one method for the given algorithm, then fall back to the code you've already written that hashes then signs. It may be better to first merge this PR, then implement the new driver dispatch at once for both sign-hash and sign-message — what do you think @ronald-cron-arm?
* [ ] `psa_exercise_key` should have cases for keys that have the `SIGN_MESSAGE` or `VERIFY_MESSAGE` usage.

**Comment by stevew817** [comment]
> It may be better to first merge this PR, then implement the new driver dispatch at once for both sign-hash and sign-message.

There's plenty of examples for functionality that has been converted to driver dispatch already. Conversion for this PR should be the same as following the footpath already created by the dispatch of `psa_sign_hash` and `psa_verify_hash`. I'd very much like for that to happen in this PR, since it's not all that much work and avoids the lead-time of yet another PR.

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2888]: @gilles-peskine-arm Shouldn't keys registered with `PSA_KEY_USAGE_SIGN_HASH` also be able to be used with `sign_message`, since `SIGN_HASH` is more permissive than `SIGN_MESSAGE`?

Asking in the interest of compatibility, i.e. keys that have so far been created with SIGN_HASH permission cannot be upgraded to SIGN_MESSAGE, so the next best thing would be to allow them to be used for the all-in-one operation too.

Comment (gilles-peskine-arm) [reply library/psa_crypto.c:2888]: Indeed, and likewise with `VERIFY`. Though this isn't the best place to implement it, since [according to the specification](https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html?highlight=psa_key_usage_sign_hash#c.PSA_KEY_USAGE_SIGN_HASH)

> if an application sets the flag PSA_KEY_USAGE_SIGN_HASH when creating a key, then the key always has the permissions conveyed by PSA_KEY_USAGE_SIGN_MESSAGE, and the flag PSA_KEY_USAGE_SIGN_MESSAGE will also be present when the application queries the usage flags of the key.

So the right place to implement this is when setting the attributes of a new key. Otherwise `psa_get_key_attributes` won't do the right thing. This is part of the scope of https://github.com/ARMmbed/mbedtls/issues/3267, but it's up to @gabor-mezei-arm whether to do it in this pull request or in a follow-up.

Upgrading existing keys in persistent storage was not in scope, so I've filed a [separate issue](https://github.com/ARMmbed/mbedtls/issues/4368) for it.

**Comment by gabor-mezei-arm** [comment]
@stevew817 @ronald-cron-arm The driver wrapper of `psa_sign_hash` is calling the crypto SE driver `p_sign` if crypto SE is enabled. Is there a need for a similar driver for the `psa_sign_message` or just use the `p_sign` for the signing step?

**Comment by ronald-cron-arm** [comment]
> 
> 
> @stevew817 @ronald-cron-arm The driver wrapper of `psa_sign_hash` is calling the crypto SE driver `p_sign` if crypto SE is enabled. Is there a need for a similar driver for the `psa_sign_message` or just use the `p_sign` for the signing step?

The current approach regarding the dynamically registered drivers is just to not break the current support, no need to add additional support. Thus here I don't think you need to do anything regarding the dynamically registered drivers.

**Comment by gabor-mezei-arm** [comment]
Rebased to development_2.x, need to be forwardported to the development branch.

File: library/psa_crypto.c
Comment (gabor-mezei-arm) [reply library/psa_crypto.c]: It will done in a separate PR.

File: tests/suites/test_suite_psa_crypto_driver_wrappers.function
Comment (stevew817) [inline tests/suites/test_suite_psa_crypto_driver_wrappers.function:64]: Pre-existing, but topical: since verification for ECDSA is the same whether the signature has been generated deterministically or not, can we remove the dependency on deterministic ECDSA here, and set the verifying algorithm to 'regular' ECDSA?

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2888]: I don't believe the blank line is according to the code standard?

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2891]: Is there a missing case here for algorithm validation on `operation == PSA_SIGN_HASH`?

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2977]: Same comments as for sign_internal.

File: library/psa_crypto_core.h
Comment (stevew817) [inline library/psa_crypto_core.h:360]: missing newline

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:3040]: Rather through `psa_driver_wrapper_sign_hash`, for those accelerators which support signing hashes but not full messages.

File: library/psa_crypto_driver_wrappers.c
Comment (stevew817) [inline library/psa_crypto_driver_wrappers.c:81]: If you make `psa_sign_message_internal` call `psa_driver_wrapper_sign_hash` instead, you can remove this block for the old-style SE interface. It's basically just unnecessary duplication like it is set up now.

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2901]: Since according to the spec linked by Gilles earlier "a key with permission `PSA_KEY_USAGE_SIGN/VERIFY_HASH` can always be used for signing/verifying a full message too, why defer that to a separate PR?

The way I see it, that would just require a minor modification to the policy checking code, and since you're introducing testing for sign_message / verify_message in this PR, I think it fully makes sense to do the policy aspect in here as well.

Comment (gabor-mezei-arm) [reply library/psa_crypto.c:2901]: The scheduling of this PR made me to put it into a separate PR. We would like to finish this PR as soon as possible.

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2878]: I'm still not quite friends with the way this logic is set up...

Correct me if I'm wrong, but the way I understand it, the logic would be:

- If operation is 'sign hash', only allow sign-hash algorithms, i.e. 'PSA_ALG_IS_SIGN_HASH( alg )`. Report an error otherwise.
- If operation is 'sign message', allow both sign-hash and sign-message algorithms, i.e. `PSA_ALG_IS_SIGN_MESSAGE( alg ) || PSA_ALG_IS_SIGN_HASH( alg )`. But since the definition of IS_SIGN_MESSAGE already accounts for the inclusion of IS_SIGN_HASH algorithms (which is honestly a bit counter-intuitive to me), that collapses to only `PSA_ALG_IS_SIGN_MESSAGE`.

And orthogonally to that, we need to verify that a signing algorithm which builds on a hash actually has a hash identifier encoded in its algorithm ID.

I believe this block (checking of algorithm value) could be written more clearly, and might even benefit from being structured as a switch-case on `operation`, since the operation type is an enum (meaning we can exhaustively address all potential values of the enum). It would also be more clear which kind of validation is performed where, since right now you're mixing validation of algorithm type v operation type, and internal algorithm type v. needed hash algorihm into a large blob.

Something like
```
switch(operation) {
  case sign_message:
    if(!PSA_ALG_IS_SIGN_MESSAGE(alg)) return error;
    break;
  case sign_hash:
    if(!PSA_ALG_IS_HASH_AND_SIGN(alg) return error;
    break;
  default:
    return error;
}

if (PSA_ALG_IS_HASH_AND_SIGN(alg) {
  // return error if algorithm ID doesn't specify which hash algo is used
}
```
would be cleaner (and more correct) IMO.

**Review by stevew817 — COMMENTED** [review]
Still waiting on feedback re. the testing of signature verification, and I have one more comment on the sign_internal / verify_internal implementation.

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2876]: Shouldn't this be PSA_SIGN_HASH instead?

File: library/psa_crypto.c
Comment (stevew817) [inline library/psa_crypto.c:2879]: Does the hash component of the hash-and-sign algorithm not need to be a valid hash algorithm when asking to sign the hash? I believe there should be a check somewhere matching the input length to the expected hash length. What do you think, @gilles-peskine-arm?

Comment (gabor-mezei-arm) [reply library/psa_crypto.c:2879]: According to the PSA specification:

> Some hash-and-sign mechanisms apply a padding or encoding to the hash. In such cases, the encoded hash must be passed to this function. The current version of this specification defines one such signature algorithm: PSA_ALG_RSA_PKCS1V15_SIGN_RAW.

It is only needed when PSA_ALG_RSA_PKCS1V15_SIGN_RAW is used.

File: tests/suites/test_suite_psa_crypto.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:2501]:
```suggestion
PSA verify message: ECDSA SECP256R1 SHA-256, good
```

File: tests/suites/test_suite_psa_crypto.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:2505]:
```suggestion
PSA verify message with keypair: ECDSA SECP256R1 SHA-256, good
```

File: tests/suites/test_suite_psa_crypto.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:2556]: Please add sign and verify test cases for `PSA_ALG_ECDSA_ANY` and `PSA_ALG_RSA_SIGN_PKCS1V15_RAW`.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: library/psa_crypto.c
Comment (stevew817) [reply library/psa_crypto.c:2868]: That makes sense to me. But who's going to replicate this change to already-existing `_internal` functions?

File: library/psa_crypto.c
Comment (stevew817) [reply library/psa_crypto.c:3011]: I'm not sure which revision this comment was left at, but as of 6eec665 I believe this is taken care of. `sign_message` calls `sign_message_internal`, which tries the `sign_message` driver. The `sign_message` driver will fallback to `sign_message_builtin` which does the hashing and calls the `sign_hash` driver.

File: library/psa_crypto.c
Comment (gilles-peskine-arm) [reply library/psa_crypto.c:2868]: If we agree on a convention, we'll just apply it to sign/verify-message here and file the rest as an issue and add it to our backlog.

File: library/psa_crypto.c
Comment (gilles-peskine-arm) [reply library/psa_crypto.c:3011]: For a transparent driver, yes (and AFAIR this was the case in every version I reviewed). But for an opaque driver, there's no second chance.

File: library/psa_crypto.c
Comment (ronald-cron-arm) [reply library/psa_crypto.c:2868]: Yes fine by me for _builtin.

File: library/psa_crypto.c
Comment (stevew817) [reply library/psa_crypto.c:3011]: Ah, I see what you mean now. That should be a super easy fix, though, so not sure it's worth pushing onto the backlog.

In `psa_sign_internal`, if the call to `psa_driver_wrapper_sign_message` returns `NOT_SUPPORTED`, proceed by calling `psa_sign_message_builtin` instead. Same for verify.

**Review by stevew817 — APPROVED** [review]
I only have minor points left, which are non-blocking if time doesn't allow for fixing them.

File: library/psa_crypto_driver_wrappers.c
Comment (stevew817) [reply library/psa_crypto_driver_wrappers.c:126]: That's true for the test driver, but not necessarily for an actual hardware driver (which might omit `sign_message` but implement `sign_hash`). The driver dispatch doc isn't clear on what happens in the auto-generated file in such a case, which I assume is done on purpose to defer these decisions until implementation time and see what would work.

I see two ways for this:

1.  Generate a case for `HW_DRIVER_OPAQUE_LOCATION` in all dispatch functions when that location value is 'known' to the autogeneration script, and throw 'NOT_SUPPORTED' and/or break from the switchcase as appropriate in order to get to the fallback.
2. Make the default case fall back to `sign_message_internal` (and reduce the potential amount of different switch cases).

I'm ambivalent, and can live with the current state of this PR, so for me this decision can be deferred until someone starts looking at implementing the driver dispatch autogeneration.

**Review by stevew817 — APPROVED** [review]
Thanks for the quick update.

**Review by stevew817 — APPROVED** [review]
Looks even better to me after these last 4 commits 👍

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by gilles-peskine-arm** [comment]
Thanks for the updates, looks good to me as well. Please make a PR for 3.0.

File: tests/suites/test_suite_psa_crypto_driver_wrappers.function
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto_driver_wrappers.function:145]: Please replace `PSA_ECC_CURVE_xxx` by `PSA_ECC_FAMILY_xxx` as well. `PSA_ECC_CURVE_xxx` are deprecated in 2.x, even if they aren't removed yet.

CI only let this through because this code isn't built in the `full` config.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

**Review by stevew817 — DISMISSED** [review]

**Review by stevew817 — APPROVED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**
